resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesExternalShuffleService.scala

```diff
@@ -0,0 +1,235 @@
+
+  // Stores a map of app id to app state (timeout value and last heartbeat)
+  private val connectedApps = new ConcurrentHashMap[String, AppState]()
+  private val indexCacheLoader = new CacheLoader[File, ShuffleIndexInformation]() {
+    override def load(file: File): ShuffleIndexInformation = new ShuffleIndexInformation(file)
+  }
+  private val shuffleIndexCache = CacheBuilder.newBuilder()
+    .maximumWeight(JavaUtils.byteStringAsBytes(indexCacheSize))
+    .weigher(new Weigher[File, ShuffleIndexInformation]() {
+      override def weigh(file: File, indexInfo: ShuffleIndexInformation): Int =
+        indexInfo.getSize
+    })
+    .build(indexCacheLoader)
+
+  private final val shuffleDir = Utils.createDirectory("/tmp", "spark-shuffle-dir")
+
+  protected override def handleMessage(
+      message: BlockTransferMessage,
+      client: TransportClient,
+      callback: RpcResponseCallback): Unit = {
+    message match {
+      case RegisterDriverParam(appId, appState) =>
+        val address = client.getSocketAddress
+        val timeout = appState.heartbeatTimeout
+        logInfo(s"Received registration request from app $appId (remote address $address, " +
+          s"heartbeat timeout $timeout ms).")
+        if (connectedApps.containsKey(appId)) {
+          logWarning(s"Received a registration request from app $appId, but it was already " +
+            s"registered")
+        }
+        val driverDir = Paths.get(shuffleDir.getAbsolutePath, appId).toFile
+        if (!driverDir.mkdir()) {
+          throw new RuntimeException(s"Failed to create dir ${driverDir.getAbsolutePath}")
+        }
+        connectedApps.put(appId, appState)
+        callback.onSuccess(ByteBuffer.allocate(0))
+
+      case Heartbeat(appId) =>
+        val address = client.getSocketAddress
+        Option(connectedApps.get(appId)) match {
+          case Some(existingAppState) =>
+            logTrace(s"Received ShuffleServiceHeartbeat from app '$appId' (remote " +
+              s"address $address).")
+            existingAppState.lastHeartbeat = System.nanoTime()
+          case None =>
+            logWarning(s"Received ShuffleServiceHeartbeat from an unknown app (remote " +
+              s"address $address, appId '$appId').")
+        }
+      case OpenParam(appId, shuffleId, mapId, partitionId) =>
+        logInfo(s"Received open param from app $appId")
+        val indexFile = getFile(appId, shuffleId, mapId, "index")
+        try {
+          val shuffleIndexInformation = shuffleIndexCache.get(indexFile)
+          val shuffleIndexRecord = shuffleIndexInformation.getIndex(partitionId)
+          val managedBuffer = new FileSegmentManagedBuffer(
+            transportConf,
+            getFile(appId, shuffleId, mapId, "data"),
+            shuffleIndexRecord.getOffset,
+            shuffleIndexRecord.getLength)
+          callback.onSuccess(managedBuffer.nioByteBuffer())
```

Comment:
we probably want to make this a stream eventually so we can send back more data?

Reply:
FileSegmentManagedBuffers already are sent in a smart way, so that you can send a large amount of data backed by a file as a stream (see `convertToNetty`).  However, the receiving end needs to know how to get the result back without just slurping it all into memory -- I left a comment elsewhere about that.